### PR TITLE
Migrate python scope "value" to new scope handler

### PR DESCRIPTION
**Pull request by saidelike**
## Checklist

- [ ] I have executed the existing tests but they seem to fail for whatever reason. Not sure why (e.g. `takeValue.yml` from a quick glance
- [x] I have tested the `visualize value` command and all the `take value` from this file and they all work fine

image

If anyone has ideas why the unit tests don't work. 

NOTES:
- I haven't included the assignment symbols like ".", "=", "/=" as I don't think they are needed when using the tree-sitter but let me know if you think otherwise.
- the "return" in the "return_statement" is something that was ignored too as not relevant anymore afaict

### CODE REVIEW

**Comment by saidelike** [comment]
And this is temporarily disabling the assignment values to see the domain of dictionary values:

image

**Review by pokey — COMMENTED** [review]
Coming along! Missing

- [ ] Domains on assignments
- [ ] Leading delimiters on everything except return values (search in scm files for `leading`)

**Comment by saidelike** [comment]
> Coming along! Missing
> 
>     * [ ]  Domains on assignments
> 
>     * [ ]  Leading delimiters on everything except return values (search in scm files for `leading`)

Ok domains on assignments done. Do you have examples for use cases where delimeters matter for the `value` scope? I understand the `,` in the `arg` scope case, but I don't see it for `value`...

**Comment by pokey** [comment]
It's if you say "chuck value". For example if you wanted to convert a dict to a set by deleting all the values, you could say "chuck every value" and it would clean up the `:`. Not totally clear how useful it is for assignments, but we support it today so I'd be inclined to preserve that behaviour 

Also, speaking of "every" you'll want to add iteration scopes as well

**Comment by saidelike** [comment]
Added value iteration scopes:

`visualize value`:

image

`visualize value iteration`:
image

File: queries/python.scm
Comment (pokey) [inline queries/python.scm:1]: It looks like you've lost a bunch of value captures by making the types more specific here. Search in [tree-sitter-python](https://github.com/tree-sitter/tree-sitter-python/blob/master/grammar.js) for `'value'`. Eg we lost [values within match statement patterns](https://www.inspiredpython.com/course/pattern-matching/mastering-structural-pattern-matching), values in [named expressions](https://realpython.com/lessons/assignment-expressions/), etc. I'd be inclined to keep a generic `value: (_)` of some sort as that is often what we want

**Review by pokey — COMMENTED** [review]
Ok I made some small tweaks in 094c70e684c5f30c4a6c4af340d5d55ecbf7206e. Note in particular how I added `!` and `!!` marks to your comments and tweaked some of the example markers underneath those lines. See https://github.com/cursorless-dev/cursorless/issues/1524 for more on that syntax

**Comment by pokey** [comment]
ok sorry thought you were sleeping so I force-pushed over my previous commit 😅

**Comment by pokey** [comment]
this PR is good to go now from my perspective; see 22bf6ea7494c75d82d653d49fcbd6f1315add155

**Comment by pokey** [comment]
all told, I made the following changes:

- Added those `!` and tweaked scope range markers as mentioned above
- Removed extraneous parens from groupings where there was no predicate
- Changed the way we do leading range to start from the end of previous named node instead of looking for `=`, `+=`, etc, as that simplified code and also fixed the failing tests that expected better whitespace cleanup
- Added some tests to catch a few other value examples

**Comment by pokey** [comment]
The last issue remaining is that the new playground file has some invalid python syntax; see the complaining pre-commit step in CI

**Comment by pokey** [comment]
We're also not supporting iteration scope for multi-value return values, eg `return a, b`. That is definitely extra credit, though, so I'd be tempted to do that in follow-up PR

**Comment by pokey** [comment]
image

File: data/playground/python/values.py [hunk 1]
```diff
@@ -37,9 +37,7 @@ def my_funk(value: str = "hello", other: bool=False) -> str:
 
 # we can say "change every value" to allow modifying all the values in one go
 def foo():
-    a = 0
-    b = 1
-    c = 2
+    pass
```
Comment (pokey) [inline data/playground/python/values.py:40]: wow looks like we have some pretty aggressive "auto-formatting" going on here 😅. cc/ @auscompgeek any idea which formatter is doing this?

File: queries/python.scm [hunk 2]
```diff
@@ -27,6 +27,58 @@
...
 
+;;!! a = 25
+;;!      ^^
+;;!   xxxxx
+;;!  ------
+(assignment
+  (_) @_.leading.start.endOf
+  .
+  right: (_) @value @_.leading.end.startOf
+) @_.domain
```
Comment (pokey) [inline queries/python.scm:38]: fwiw I do think this pattern and the following one (`augmented_assignment`) are potentially a compelling use case for `#has-type?` predicate, as it would avoid duplication. @AndreasArvidsson wdyt?

File: queries/python.scm [hunk 3]
```diff
@@ -27,6 +27,58 @@
...
+  (#not-type? @_.domain subscript)
+)
+
+;;!! return 1
+;;!         ^
+;;!        xx
+;;!  --------
+;;
+;; NOTE: in tree-sitter, both "return" and the "1" are children of `return_statement`
+;; but "return" is anonymous whereas "1" is named node, so no need to exclude explicitly
```
Comment (pokey) [inline queries/python.scm:77]: This comment might be overkill. We leverage this kind of assumption all over the place. For example, I'm leveraging this kind of assumption in my leading matchers above, where I rely on the fact that `=` is anonymous whereas the name / type before it is not

**Comment by saidelike** [comment]
> notice how the removal range extends all the way past the space before the `=` etc
> 
> image
> 

Looks great. What command can you say to see the removal range?

**Comment by pokey** [comment]
"visualize value removal"

see [docs for visualizer](https://www.cursorless.org/docs/user/scope-visualizer/), although at this point I think you now know everything in that doc 😄

File: queries/python.scm (hunk 2 above)
Comment (saidelike) [reply queries/python.scm:38]: To me it is the same as dictionary vs dictionary_comprehension as discussed in https://github.com/cursorless-dev/cursorless/commit/36c2f4c7e5227086cf9ed098a5c8d787fea063d5

Comment (pokey) [reply queries/python.scm:38]: Agreed it is quite similar. In that case we managed to avoid duplication another way, but I don't see a way out of it here other than `has-type`

**Comment by pokey** [comment]
> The last issue remaining is that the new playground file has some invalid python syntax; see the complaining pre-commit step in CI

Once this is fixed I think we're good to merge

File: data/playground/python/values.py (hunk 1 above)
Comment (auscompgeek) [reply data/playground/python/values.py:40]: It's probably ruff's pyflakes (F) autofixes yeeting the unused variables, then the unused literal expressions?

Comment (josharian) [reply data/playground/python/values.py:40]: I guess I would prefer a blanket exception for everything in the playground. I have several things in the playground that are not gofmt compliant (but are valid go code ) specifically so that I can experiment to see whether cursorless does the right thing with them.

Comment (pokey) [reply data/playground/python/values.py:40]: I am very much on the fence, but lean slightly towards formatting playground but being liberal with comments to disable auto-formatter / linter where we want. @josharian is it not easy to disable gofmt on a per-line basis?

I don't feel strongly, tho. I could see the benefit to having full freedom in the playground

Comment (josharian) [reply data/playground/python/values.py:40]: > is it not easy to disable gofmt on a per-line basis

no, by design gofmt is impossible to selectively disable and has no knobs whatsoever

(Happy to expound upon the thinking behind those design decisions over voice, if desired)

Comment (josharian) [reply data/playground/python/values.py:40]: However, there are also no pre-commit hooks for Go at all, so far this is not a problem in the playground or anywhere. :)

Comment (pokey) [reply data/playground/python/values.py:40]: I could see a case for blanket disabling formatter in playground. That being said, the formatter has detected real syntax errors in these new python playground files, so it is providing some value here

Comment (josharian) [reply data/playground/python/values.py:40]: We could have the pre-commit step check for the actual property we want, namely valid syntax

Comment (josharian) [reply data/playground/python/values.py:40]: It depends in part on how active the formatter is. There is some danger of writing queries that only match when using the same formatter as us, particularly if the formatter is not whitespace only. 🤷🏽

Comment (pokey) [reply data/playground/python/values.py:40]: Yeah that's a good point. I think maximal flexibility is not necessarily a bad thing in the playground. Interesting idea to just check for valid syntax. Altho tbh we even might want to violate that in some cases because error tolerance is one of the reasons we like tree-sitter

Comment (pokey) [reply data/playground/python/values.py]: I'm convinced. See 43939f76e080c3000e9d23df27b1004fe667f359. Any objections @auscompgeek @josharian @AndreasArvidsson ?

Comment (AndreasArvidsson) [reply data/playground/python/values.py]: No sounds resonable

**Review by pokey — APPROVED** [review]
Looks good to me!
